### PR TITLE
Cimg lockstep

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/graphics/gmic/default.nix
Comment (lilyinstarlight) [inline pkgs/tools/graphics/gmic/default.nix:73]: This will make it more difficult to override because overriding `version` still requires this hash on `gmic_stdlib` to be changed (and before `gmic.overrideAttrs (oldAttrs: { version = ...; src = ...; gmic_stdlib = ...; })` would have been an option)

Maybe the derivation should just use `srcs` instead?

**Review by lilyinstarlight — COMMENTED** [review]
These changes mostly look fine enough to me and brief testing indicates everything is working, thank you!

I left one question since this does make it so that `gmic.gmic_stdlib` is no longer a thing

---

Result of `nixpkgs-review pr 228986` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build (also failing in Hydra):</summary>
  <ul>
    <li>openroad (<a href="https://hydra.nixos.org/build/217586470">Hydra failure</a>)</li>
  </ul>
</details>
<details>
  <summary>26 packages built:</summary>
  <ul>
    <li>ceph (ceph-dev)</li>
    <li>ceph-client</li>
    <li>ceph-csi</li>
    <li>ceph.dev (ceph-dev.dev)</li>
    <li>ceph.doc (ceph-dev.doc)</li>
    <li>libceph (ceph.lib ,libceph.dev ,libceph.doc ,libceph.lib ,libceph.man)</li>
    <li>ceph.man (ceph-dev.man)</li>
    <li>cimg</li>
    <li>cimg.doc</li>
    <li>darktable</li>
    <li>gimp-with-plugins</li>
    <li>gimpPlugins.gmic</li>
    <li>gmic</li>
    <li>gmic-qt</li>
    <li>gmic.dev</li>
    <li>gmic.lib</li>
    <li>gmic.man</li>
    <li>imgcat</li>
    <li>phash</li>
    <li>photoprism</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>sambaFull (samba4Full)</li>
    <li>sambaFull.dev (samba4Full.dev)</li>
    <li>sambaFull.man (samba4Full.man)</li>
  </ul>
</details>

File: pkgs/tools/graphics/gmic/default.nix
Comment (AndersonTorres) [reply pkgs/tools/graphics/gmic/default.nix:73]: You are right. I will "revert" this.

Well, we can investigate how this file can be generated from sources to begin with. Reading the Makefile, it looks like this file can be generated by calling another directive.

Comment (AndersonTorres) [reply pkgs/tools/graphics/gmic/default.nix]: Oh frak, this is self-referent! It uses gmic to generate gmic-stdlib!

https://github.com/GreycLab/gmic/blob/cc41e1d5de0fd01f6ae48d5d8df6fc73a9188558/src/Makefile#L735

I need to re-read all theory about bootstrap in Nixpkgs...

Comment (lilyinstarlight) [reply pkgs/tools/graphics/gmic/default.nix]: > Oh frak, this is self-referent! It uses gmic to generate gmic-stdlib!

Yep, I noticed that too when I last looked at it...

We only started needing to download it separately when we started needing to pull the main sources from GitHub (since the tarball actually has the file already in it). We had to pull from GitHub since the CMake files stopped being included with the tarball a few versions back (#179646)

We could perhaps switch to the Makefile-based G'MIC build which is included in the tarball, but I'm not sure we can get G'MIC-Qt to find that properly without the generated CMake file...

It's a bit of a mess tbh so downloading is probably best for now

Comment (lilyinstarlight) [reply pkgs/tools/graphics/gmic/default.nix]: Also I think it may have broken the updateScript when it wasn't an attr on the derivation anymore, but I'd have to double check to be sure

Either way, just pulling it back out right now looks good, thanks!

File: pkgs/tools/graphics/gmic-qt/default.nix
Comment (lilyinstarlight) [inline pkgs/tools/graphics/gmic-qt/default.nix:85]: Actually, I have one more question. Why was `sourceRoot` removed in favor of this?

Afaik we shouldn't need anything from the root dir of the tarball, and I thought `sourceRoot` was the right way to indicate that everything will be happening from a subdir

Comment (AndersonTorres) [reply pkgs/tools/graphics/gmic-qt/default.nix:85]: It is a bit subjective. 

`sourceRoot` is not reliable, because it depends on the previous `sourceRoot`, and it is opaquely determined by the fetcher.
Some like fetchurl create `${pname}-${version}`, others like fetchzip use `source`. And any future change on the internals of those fetchers can break this.

Indeed a similar thing occurred a short time ago: many `data/fonts` changed from using `fetchzip` internal machinery to using the good old `stdenv`, because `fetchzip` changed internally.

Because of it, I prefer to `cd` to the desired directory instead.

Comment (lilyinstarlight) [reply pkgs/tools/graphics/gmic-qt/default.nix:85]: That makes sense, but sucks a lot. I wish we had something that _did_ basically do a `cd` then without having to know what the fetcher does. Doing `cd` in `preConfigure` instead of `sourceRoot` will fail some builders (but I imagine `postPatch` should work pretty generally -- obviously that's semantically fraught though)

I guess that's out of scope for this PR though since this works 👍🏻

**Review by lilyinstarlight — APPROVED** [review]
Looks good to me!

Paging @jtojnar if they had any comments about any of the changes, since they wrote most of this at some point

Also I did play around with bootstrapping `gmic_stdlib.h` and it's totally doable by compiling with a fake, mostly empty `gmic_stdlib.h` and then just doing `ninjaBuildPhase` twice, once before and once after generating the file. The complexity it adds is almost definitely not worth it though, so I still think the additional fetch is best for now (either until upstream decides to start adding the CMake files back to the tarballs or we decide on some other jank to replace it)

**Comment by jtojnar** [comment]
> Paging @jtojnar if they had any comments about any of the changes, since they wrote most of this at some point

I do not see other changes than bump, `finalAttrs` switch and attribute reordering. Ideally, for messy diffs like this, there should either be a one commit per change, or the changes should at least be listed in the commit message so that reviewer knows they did not miss anything.

**Comment by lilyinstarlight** [comment]
> I do not see other changes than bump and some trivial formatting changes (`finalAttrs` switch and attribute reordering).

I should have been clearer, I just wanted a second pair of eyes to confirm that there were no functional changes :)

(Also this PR originally moved `gmic_stdlib` into a `let ... in` but that has since been changed back since it would mess up some stuff and make overriding version very difficult)

I'll go ahead and merge if you didn't have any other comments

**Comment by AndersonTorres** [comment]
Well, I can try to split the cosmetic changes. The only problem is the redundant waiting for ofborg CI.

**Comment by lilyinstarlight** [comment]
> Well, I can try to split the cosmetic changes. The only problem is the redundant waiting for ofborg CI.

I'm mostly indifferent to splitting out the cosmetic changes at this point. It would have made it a much easier review if they were split out from the start, and it would probably be better to do that in future PRs, so that it is easier to pull apart what's a functional change and what's not. But I've also already reviewed this as-is and confirmed things work and it's not *that* horrible of a diff

Since they brought it up, I'll leave it to @jtojnar if they would rather it be split out or are fine to let this get merged (if needed, I can even just split it out myself if you don't want to/don't have the time to and push to the PR myself)

**Review by lilyinstarlight — APPROVED** [review]
The diff still looks good and it looks like everything still works in testing

I went ahead and confirmed cosmetic changes were in fact cosmetic and did not result in a meaningfully different derivation (other than ordering or noticeable changes like `sourceRoot` -> `preConfigure`)

Thank you again @AndersonTorres!

**Comment by risicle** [comment]
(`nixpkgs-review` revealed no new failures, macos 10.15, didn't finish nixos x86_64...)

**Comment by lilyinstarlight** [comment]
> (`nixpkgs-review` revealed no new failures, macos 10.15, didn't finish nixos x86_64...)

(I didn't post it, but I did run a nixpkgs-review on x86_64-linux before merging that passed. It's unfortunate ceph+qemu are reverse dependencies now...)

**Pull request merged**
